### PR TITLE
capacity providers

### PR DESCRIPTION
**Pull request by chathsuom**

### CODE REVIEW

File: src/main/java/com/cloudbees/jenkins/plugins/amazonecs/ECSService.java
Comment (Chris9224) [inline src/main/java/com/cloudbees/jenkins/plugins/amazonecs/ECSService.java:310]: Should we use a constant for "FARGATE" and "awsvpc"

**Review by Chris9224 — APPROVED** [review]

File: src/main/java/com/cloudbees/jenkins/plugins/amazonecs/ECSService.java
Comment (chathsuom) [reply src/main/java/com/cloudbees/jenkins/plugins/amazonecs/ECSService.java:310]: Not sure how to do that properly. (not from a development background). Do you have a sample somewhere in the code, so that I can refer.

Comment (Chris9224) [reply src/main/java/com/cloudbees/jenkins/plugins/amazonecs/ECSService.java:310]: You can re-use the enum in com.amazonaws.services.ecs.model
Example, for "FARGATE" you can use the enum in com.amazonaws.services.ecs.model.LaunchType
And for "awsvpc", you can use the enum in  com.amazonaws.services.ecs.model.NetworkMode

**Comment by Xantier** [comment]
Merging this in adds a new permission requirement. Should probably add that to the README file.
`ecs:DescribeClusters` on the main Jenkins instance is needed.

File: src/main/java/com/cloudbees/jenkins/plugins/amazonecs/ECSService.java
Comment (chathsuom) [reply src/main/java/com/cloudbees/jenkins/plugins/amazonecs/ECSService.java]: Fixed

**Comment by chathsuom** [comment]
> Merging this in adds a new permission requirement. Should probably add that to the README file.
> `ecs:DescribeClusters` on the main Jenkins instance is needed.

Fixed

**Comment by Xantier** [comment]
@webratz This could probably be something to merge in. We've been running this on our boxes for a while now and both Fargate and EC2 based tasks handle this nicely with the default capacity provider setting.

**Comment by MikeKroell** [comment]
Do we know when this will be officially merged in?

**Comment by webratz** [comment]
see #200

**Comment by 9970807747** [comment]
@webratz We started using this plugin recently and realized that you cannot cap max nodes it can spin up. Is there anything holding up this PR to be merged? This is one of the important feature for us to use this plugin. I would really appreciate if this can be merged.

**Comment by willychase** [comment]
@webratz , If it isn't too much to ask could this be merged? I understand you are looking for a new maintainer, but getting this change in would allow my Org to use a non-forked version of this project.

**Comment by tdspencer3** [comment]
We would really like for this PR to be merged in. We are actively using this code in our production environment (going on 2 months now). But it means we had to build it and side-load it because the code has not been released in this plugin yet.

**Comment by gokhanoner** [comment]
@willychase @tdspencer3 I'll look into this this week & merge ( I'm also using a similar version in my company as well, for the last 4 months now)

**Comment by carlosrodf** [comment]
Any plans to merge this soon? this would be awesome to have without having to custom build/install it

**Comment by markyjackson-taulia** [comment]
@carlosrodf Let me see if I can get CI to pass and I will review and merge

**Comment by markyjackson-taulia** [comment]
@carlosrodf Wheels fall of the bus here:
```
[2021-01-28T23:52:20.636Z] Caused by: org.apache.maven.plugin.compiler.CompilationFailureException: Compilation failure
[2021-01-28T23:52:20.636Z] /home/jenkins/workspace/Plugins_amazon-ecs-plugin_PR-211/src/main/java/com/cloudbees/jenkins/plugins/amazonecs/ECSTaskTemplate.java:[1038,41] no suitable constructor found for ECSService(java.lang.String,java.lang.String)
[2021-01-28T23:52:20.636Z]     constructor com.cloudbees.jenkins.plugins.amazonecs.ECSService.ECSService(java.lang.String,java.lang.String,java.lang.String) is not applicable
[2021-01-28T23:52:20.636Z]       (actual and formal argument lists differ in length)
[2021-01-28T23:52:20.636Z]     constructor com.cloudbees.jenkins.plugins.amazonecs.ECSService.ECSService(java.util.function.Supplier<com.amazonaws.services.ecs.AmazonECS>) is not applicable
[2021-01-28T23:52:20.636Z]       (actual and formal argument lists differ in length)
[2021-01-28T23:52:20.636Z] 
```
@carlosrodf, @chathsuom will need to correct that.

**Review by markyjackson-taulia — CHANGES_REQUESTED** [review]
Seeing failures in
```
[2021-01-28T23:52:20.636Z] Caused by: org.apache.maven.plugin.compiler.CompilationFailureException: Compilation failure
[2021-01-28T23:52:20.636Z] /home/jenkins/workspace/Plugins_amazon-ecs-plugin_PR-211/src/main/java/com/cloudbees/jenkins/plugins/amazonecs/ECSTaskTemplate.java:[1038,41] no suitable constructor found for ECSService(java.lang.String,java.lang.String)
[2021-01-28T23:52:20.636Z]     constructor com.cloudbees.jenkins.plugins.amazonecs.ECSService.ECSService(java.lang.String,java.lang.String,java.lang.String) is not applicable
[2021-01-28T23:52:20.636Z]       (actual and formal argument lists differ in length)
[2021-01-28T23:52:20.636Z]     constructor com.cloudbees.jenkins.plugins.amazonecs.ECSService.ECSService(java.util.function.Supplier<com.amazonaws.services.ecs.AmazonECS>) is not applicable
[2021-01-28T23:52:20.636Z]       (actual and formal argument lists differ in length)
[2021-01-28T23:52:20.636Z] 
```

**Comment by chathsuom** [comment]
> Seeing failures in
> 
> ```
> [2021-01-28T23:52:20.636Z] Caused by: org.apache.maven.plugin.compiler.CompilationFailureException: Compilation failure
> [2021-01-28T23:52:20.636Z] /home/jenkins/workspace/Plugins_amazon-ecs-plugin_PR-211/src/main/java/com/cloudbees/jenkins/plugins/amazonecs/ECSTaskTemplate.java:[1038,41] no suitable constructor found for ECSService(java.lang.String,java.lang.String)
> [2021-01-28T23:52:20.636Z]     constructor com.cloudbees.jenkins.plugins.amazonecs.ECSService.ECSService(java.lang.String,java.lang.String,java.lang.String) is not applicable
> [2021-01-28T23:52:20.636Z]       (actual and formal argument lists differ in length)
> [2021-01-28T23:52:20.636Z]     constructor com.cloudbees.jenkins.plugins.amazonecs.ECSService.ECSService(java.util.function.Supplier<com.amazonaws.services.ecs.AmazonECS>) is not applicable
> [2021-01-28T23:52:20.636Z]       (actual and formal argument lists differ in length)
> [2021-01-28T23:52:20.636Z] 
> ```

I will have a look soon.

**Comment by chathsuom** [comment]
> Seeing failures in
> 
> ```
> [2021-01-28T23:52:20.636Z] Caused by: org.apache.maven.plugin.compiler.CompilationFailureException: Compilation failure
> [2021-01-28T23:52:20.636Z] /home/jenkins/workspace/Plugins_amazon-ecs-plugin_PR-211/src/main/java/com/cloudbees/jenkins/plugins/amazonecs/ECSTaskTemplate.java:[1038,41] no suitable constructor found for ECSService(java.lang.String,java.lang.String)
> [2021-01-28T23:52:20.636Z]     constructor com.cloudbees.jenkins.plugins.amazonecs.ECSService.ECSService(java.lang.String,java.lang.String,java.lang.String) is not applicable
> [2021-01-28T23:52:20.636Z]       (actual and formal argument lists differ in length)
> [2021-01-28T23:52:20.636Z]     constructor com.cloudbees.jenkins.plugins.amazonecs.ECSService.ECSService(java.util.function.Supplier<com.amazonaws.services.ecs.AmazonECS>) is not applicable
> [2021-01-28T23:52:20.636Z]       (actual and formal argument lists differ in length)
> [2021-01-28T23:52:20.636Z] 
> ```

Fixed

**Review by markyjackson-taulia — APPROVED** [review]
/lgtm

**Comment by carlosrodf** [comment]
@markyjackson-taulia can an official release be made that includes this?

**Pull request merged**
